Source/WebCore/Modules/url-pattern/URLPatternCanonical.cpp

```diff
@@ -0,0 +1,257 @@
+ExceptionOr<String> canonicalizeIPv6Host(const String& value, BaseURLStringType valueType)
+{
+    if (valueType == BaseURLStringType::Pattern)
+        return String { value };
+
+    StringBuilder result;
+    result.reserveCapacity(result.length());
+
+    for (auto codepoint : StringView(value).codePoints()) {
+        if (isInvalidIPv6HostCodePoint(codepoint))
+            return Exception { ExceptionCode::TypeError, "Invalid input to canonicalize a URL IPv6 host string."_s };
+
+        result.append(toASCIILower(codepoint));
+    }
+
+    return result.toString();
+}
+
+// https://urlpattern.spec.whatwg.org/#canonicalize-a-port, combined with https://urlpattern.spec.whatwg.org/#process-port-for-init
+ExceptionOr<String> canonicalizePort(const String& portValue, std::optional<StringView> protocolValue, BaseURLStringType portValueType)
+{
+    if (portValue.isEmpty())
+        return String { portValue };
+
+    if (portValueType == BaseURLStringType::Pattern)
+        return String { portValue };
+
+    URL dummyURL("https://www.webkit.org"_s);
+
+    if (protocolValue.has_value())
+        dummyURL.setProtocol(protocolValue.value());
+
+    dummyURL.setPort(parseInteger<uint16_t>(portValue));
+
+    if (!dummyURL.isValid())
+        return Exception { ExceptionCode::TypeError, "Invalid input to canonicalize a URL port string."_s };
+
+    return String::number(dummyURL.port().value());
+}
+
+// https://urlpattern.spec.whatwg.org/#canonicalize-an-opaque-pathname
+static ExceptionOr<String> canonicalizeOpaquePath(const String& value)
+{
+    URL dummyURL("https://www.webkit.org"_s);
+    dummyURL.setPath(value);
+
+    if (!dummyURL.isValid())
+        return Exception { ExceptionCode::TypeError, "Invalid input to canonicalize a URL opaque path string."_s };
+
+    return String { value };
+}
+
+// https://urlpattern.spec.whatwg.org/#canonicalize-a-pathname, combined with https://urlpattern.spec.whatwg.org/#process-pathname-for-init
+ExceptionOr<String> canonicalizePath(const String& pathnameValue, StringView protocolValue, BaseURLStringType pathnameValueType)
+{
+    if (pathnameValue.isEmpty())
+        return String { pathnameValue };
+
+    if (pathnameValueType == BaseURLStringType::Pattern)
+        return String { pathnameValue };
+
+    if (protocolValue == "ftp"_s || protocolValue == "file"_s
+        || protocolValue == "http"_s || protocolValue == "https"_s
+        || protocolValue == "ws"_s || protocolValue == "wss"_s) {
+
+        bool hasLeadingSlash = pathnameValue.startsWith('/');
+        auto modifiedValue = hasLeadingSlash ? ""_s : "/-"_s;
+
+        auto maybeAddSlashPrefix = makeString(modifiedValue, pathnameValue);
+
+        URL dummyURL("https://www.webkit.org"_s);
+        dummyURL.setPath(maybeAddSlashPrefix);
+
+        if (!dummyURL.isValid())
+            return Exception { ExceptionCode::TypeError, "Invalid input to canonicalize a URL path string."_s };
+
+        auto result = dummyURL.path().toString();
```

Comment:
Very sad for perf to convert to a String just to be calling substring below and re-allocating yet another string. This would be more efficient:
```
auto result = dummyURL.path();
if (hasLeadingSlash)
    result = result.substring(2);
return result.toString();
```